reactos/drivers/usb/usbhub_new/usbhub.c

```diff
@@ -0,0 +1,5074 @@
+
+    DescriptorLen = MAXUCHAR;
+
+    while (TRUE)
+    {
+        ConfigDescriptor = ExAllocatePoolWithTag(NonPagedPool,
+                                                 DescriptorLen,
+                                                 USB_HUB_TAG);
+
+        if (!ConfigDescriptor)
+        {
+            Status = STATUS_INSUFFICIENT_RESOURCES;
+            break;
+        }
+
+        Status = USBH_SyncGetDeviceConfigurationDescriptor(DeviceObject,
+                                                           ConfigDescriptor,
+                                                           DescriptorLen,
+                                                           &ReturnedLen);
+
+        if (ReturnedLen < sizeof(USB_CONFIGURATION_DESCRIPTOR))
+        {
+            Status = STATUS_DEVICE_DATA_ERROR;
+        }
+
+        if (!NT_SUCCESS(Status))
+        {
+            break;
+        }
+
+        *OutDescriptor = ConfigDescriptor;
+
+        if (ConfigDescriptor->wTotalLength <= DescriptorLen)
+        {
+            break;
+        }
+
+        DescriptorLen = ConfigDescriptor->wTotalLength;
+
+        ExFreePool(ConfigDescriptor);
+        *OutDescriptor = NULL;
+    }
+
+    if (NT_SUCCESS(Status))
+    {
+        if (ReturnedLen < ConfigDescriptor->wTotalLength)
+        {
+            Status = STATUS_DEVICE_DATA_ERROR;
+        }
+    }
+    else
+    {
+        if (ConfigDescriptor)
+        {
+            ExFreePool(ConfigDescriptor);
+        }
+
+        *OutDescriptor = NULL;
+    }
+
+    return Status;
+}
+
+NTSTATUS
+NTAPI
+USBH_SyncGetHubDescriptor(IN PUSBHUB_FDO_EXTENSION HubExtension)
+{
+    PUSB_EXTHUB_INFORMATION_0 ExtendedHubInfo;
+    ULONG NumberPorts;
+    PUSBHUB_PORT_DATA PortData;
+    USHORT RequestValue;
+    ULONG NumberOfBytes;
+    NTSTATUS Status;
+    PUSB_HUB_DESCRIPTOR HubDescriptor = NULL;
+    ULONG ix;
+
+    DPRINT("USBH_SyncGetHubDescriptor: ... \n");
+
+    ExtendedHubInfo = ExAllocatePoolWithTag(NonPagedPool,
+                                            sizeof(USB_EXTHUB_INFORMATION_0),
+                                            USB_HUB_TAG);
+
+    if (!ExtendedHubInfo)
+    {
+        Status = STATUS_INSUFFICIENT_RESOURCES;
+        goto ErrorExit;
+    }
+
+    RtlZeroMemory(ExtendedHubInfo, sizeof(USB_EXTHUB_INFORMATION_0));
+
+    Status = USBHUB_GetExtendedHubInfo(HubExtension, ExtendedHubInfo);
+
+    if (!NT_SUCCESS(Status))
+    {
+        ExFreePoolWithTag(ExtendedHubInfo, USB_HUB_TAG);
+        ExtendedHubInfo = NULL;
+    }
+
+    NumberOfBytes = sizeof(USB_HUB_DESCRIPTOR);
+
+    HubDescriptor = ExAllocatePoolWithTag(NonPagedPool,
+                                          NumberOfBytes,
+                                          USB_HUB_TAG);
+
+    if (!HubDescriptor)
+    {
+        Status = STATUS_INSUFFICIENT_RESOURCES;
+        goto ErrorExit;
+    }
+
+    RtlZeroMemory(HubDescriptor, NumberOfBytes);
+
+    RequestValue = 0;
+
+    while (TRUE)
+    {
+        while (TRUE)
```

Comment:
usbhub.c:1089: this looks like it could easily become an infinite loop. It should only ever run twice (once with 0 and once with 0x2900), right? Should there be another exit condition?